src/handlers/strategies/directoryListing.ts

```diff
@@ -111,41 +167,61 @@ export async function listDirectory(
   env: Env
 ): Promise<Response> {
   const delimitedPrefixes = new Set<string>();
-  const objects: R2Object[] = [];
+  const objects: _Object[] = []; // s3 sdk types are weird
+
+  // Create an S3 client instance to interact with the bucket.
+  // There is a limit in the size of the response that
+  //  a binding can return. We kept hitting it due to the
+  //  size of our paths, causing us to send a lot of requests
+  //  to R2 which in turn added a lot of latency. The S3 api
+  //  doesn't have that response body size constraint so we're
+  //  using it for now.
+  const client = new S3Client({
+    region: 'auto',
+    endpoint: env.S3_ENDPOINT,
+    credentials: {
+      accessKeyId: env.S3_ACCESS_KEY_ID,
+      secretAccessKey: env.S3_ACCESS_KEY_SECRET,
+    },
+  });
 
   let truncated = true;
   let cursor: string | undefined;
 
   while (truncated) {
-    const result = await env.R2_BUCKET.list({
-      prefix: bucketPath,
-      delimiter: '/',
+    const result: ListObjectsV2CommandOutput = await fetchR2Result(
+      client,
+      bucketPath,
       cursor,
-    });
+      env
+    );
 
     // R2 sends us back the absolute path of the object, cut it
-    result.delimitedPrefixes.forEach(prefix =>
-      delimitedPrefixes.add(prefix.substring(bucketPath.length))
-    );
+    result.CommonPrefixes?.forEach(path => {
+      if (path.Prefix !== undefined)
+        delimitedPrefixes.add(path.Prefix.substring(bucketPath.length));
+    });
 
-    const hasIndexFile = result.objects.find(object =>
-      object.key.endsWith('index.html')
+    const hasIndexFile = result.Contents?.find(
```

Comment:
If there's a `?` meaning that `hasIndexFile` can be of type `null`.

I rather do a `result.Contents ? .... : false`

Reply:
Also feels kinda weird that `hasIndexFile` is a find. If you just want to check if file exists.

You should be using `results.Contents.some`, and ensure that `hasIndexFile` is  Boolean either by `!!` or `Boolean()`